### DISCUSSION TITLE
Deploying Authelia on Kubernetes

### DISCUSSION THREAD

**Discussion by jefke24** [authelia/authelia#5399] [Q&A] [answered]
Hi All,

I'm hoping someone here can point me in the right direction, been banging my head against the wall for a couple of days now.

My setup is as follows:
Deploying Authelia and Nextcloud on kubernetes.

Authelia is deployed in namespace "authelia", nextcloud is deployed in namespace "nextcloud".
(deploying on K3S, if that would be relevant).

I'm exposing both via nginx ingress, can connect to each separately without any issues.
Authelia is accessible in my internal network via auth.home.myinternalnetwork.org, nextcloud is accessible via nextcloud.home.myinternalnetwork.org.

All good up until this point.
Now I want to do OIDC-based authentication for Nextcloud against Authelia.
So I'm not using the nginx ingress integration, I just want to configure Nextcloud as an OIDC client with client id and client secret against Authelia.

I'm using the official OIDC pluging from nextcloud itself.
In nextcloud, in this plugin configuration, I configure:
- the client id 
- the client secret
- the well-known OIDC endpoint of authelia

And this last one is where I run into issues.

Nextcloud and Authelia are in separate namespaces. Nextcloud needs to be able to reach the well-known OIDC endpoint.
Therefore, in the nextcloud namespace, I deployed the following service:

`kind: "Service"
apiVersion: "v1"
metadata:
  name: "authelia-service"
spec:
  type: ExternalName
  externalName: authelia.authelia.svc.cluster.local`

I configure the endpoint in nextcloud as:
http://authelia-service/.well-known/openid-configuration

So, when I execute a shell in my nextcloud pod, I can successfully curl that URL, it gives me the openid-configuration.
However, I noticed that this endpoint takes over the host-header it receives to fill in all the URL's.

So the result is, when I go to the nextcloud login page and select the authelia OIDC provider, my browser gets redirected to
https://authelia-service/<additional_params>, which is obviously not reachable from outside the cluster.

Apparently when using a service with externalName, the host-header sent is the same as the service-name (authelia-service in this case), and the Authelia OIDC endpoint takes over this host header to replace/fill-in all those URL's.

So how can I get around this, and make sure that nextcloud can connect to the authelia service in the other namespace BUT also redirects the browser to the correct external endpoint (auth.home.myinternalnetwork.org)?

Also:
- I'm accessing authelia.authelia.svc.cluster.local over port 80, since I'm doing HTTPS termination on the nginx ingress controller via auth.home.myinternalnetwork.org
- I'm deploying Authelia using the latest version of the helm chart.

I hope someone here can point me in the right direction.

**Comment by james-d-elliott** [comment]
In order for this metadata to be corerect you must either:

1. Use the fully qualified URL and appropriately configure DNS (recommended as OIDC should only operate over https)
2. Appropriately set the [documented required headers](https://www.authelia.com/integration/proxies/introduction/#required-headers).  There are however a few notes as this is effectively a hack around the issue:
    1. When the service attempts to exchange authorization codes for tokens it will do this over the back-channel in the standard flows as per the  spec. This means regardless the application *MUST* be able to make  requests to the fully qualified external URL (https).

**Reply by james-d-elliott** [reply]
While we could *technically* hard code this into the config, at this stage we have intentionally not done this. This is because we have implemented multi-domain support for session cookies and intend to extend this to OIDC with multi-issuer deployments (one per domain). This will require us to match the requested URL to a a specific issuer, and 404 otherwise. We could per-emptively do this for single-issuer configs but  this will likely break more users than it fixes.

**Reply by jefke24** [reply]
Ok, so basically you're stating that the only way to get this to work, is to force nextcloud to connect to https://auth.home.myinternalnetwork.org, instead of auth-service or authelia.authelia.svc.cluster.local?

I already had a feeling that would be needed.
But that basically means that I would need to force those requests to go out from the nextcloud namespace, make a round-trip outside and come back in via the ingress linked to auth.home.myinternalnetwork.org?

The problem is, I have no idea how I can achieve this, I thought that the only way to do this from the nextcloud namespace is using an externalname-service, but that will never work...

**Reply by james-d-elliott** [reply, marked as answer]
There are multiple ways to skin this particular cat. But the main requirement is the application must be able to make the request over `https`. This  one is not optional as it's a strict requirement due to OAuth 2.0 and OpenID Connect 1.0 standards which are  beyond our control. If  it's not required by every OAuth 2.0 and OpenID Connect 1.0 endpoint it's required by a majority of them.

As far as how you solve your issue you can either configure some form of local DNS to handle this, for example maybe [configure the Pods hosts file](https://kubernetes.io/docs/tasks/network/customize-hosts-file-for-pods/#adding-additional-entries-with-hostaliases) or the in-cluster DNS to point to the proxy (as the proxy should do TLS termination), or configure LAN DNS which points to the same.  Otherwise NAT hairpin may be an avenue that you can explore.

**Reply by jefke24** [reply]
When you talk about the in-cluster DNS, since I'm using K3S with coredns, are you referring to something like this?
https://coredns.io/plugins/forward/

So basically add a specific forwarder which would point *.home.myinternalnetwork.org to the IP my internal DNS running in my local LAN?

So if my internal DNS in my lan would be running on 192.168.1.1, something like this?

`
home.myinternalnetwork.org {
    forward . 192.168.1.1
    log
}

. {
    forward . /etc/resolv.conf
    log
}
`

**Reply by james-d-elliott** [reply]
CoreDNS is a bit beyond what I can specifically help with but I found this:

https://coredns.io/2017/05/08/custom-dns-entries-for-kubernetes/

Basically you want to configure a DNS entry that allows pods from within your cluster to properly connect to the proxy to  handle these requests.

**Reply by jefke24** [reply]
Hmmmm, thanks, I'm going to have a look at this later today and update here if I get this running, or not :)
Might then help the next victim like me who runs into this :)

Oh, and by the way, already many thanks for helping me get my head around this, or at least pointed in the right direction!

**Reply by james-d-elliott** [reply]
Yeah let me know how you go..there is a way technically to set the FQDN of the nginx pod but that seems like a not very useful idea. I have also asked in a few places about a flavor agnostic way, but no replies yet. Would be good to add a FAQ if there is something k8s universal (CoreDNS isn't unfortunately).

**Reply by jefke24** [reply]
Alright, finally got it to work.
Didn't need to use this rewrite functionality in the end.

But thanks to your input, I started looking in all the right places.
Problem was: I forgot that I was temporarily testing authelia and nextcloud using adjustments in my local /etc/hosts on my laptop.

However, the auth domain was also still present on my local internal DNS, which was being used by nextcloud for the redirect on kubernetes but... it was on the DNS-server level still using an old A-record pointing to an IP which wasn't being used anymore...
Hence the hostname via DNS from the kubernetes cluster never resolved.

The only thing I'm seeing now it seems is that, every time I authenticate in Authelia, I get asked again and again for consent for sharing certain info with nextcloud after authenticating before being forwarded to nextcloud.

So authenticating works now, but it seems as if I get presented with a consent request on EVERY single authentication.
No idea if this is a problem with Authelia or Nextcloud.

**Reply by james-d-elliott** [reply]
Consent should generally be requested on each authorization unless either the user choses to remember the consent or the app was granted refresh privileges and refreshes the tokens. See https://www.authelia.com/configuration/identity-providers/open-id-connect/#consent_mode

**Reply by jefke24** [reply]
Excellent, that fixed it.
You are a god, all working now, thanks!

Glad I got it working, was looking for a good alternative to Keycloak, just to get rid of my dependency on Redhat/IBM products.
Check that box :)
